src/com/linkedin/parseq/BaseTask.java

```diff
@@ -389,18 +391,24 @@ private void appendTaskStackTrace(final Throwable error) {
     error.setStackTrace(concatenatedStackTrace);
   }
 
-  protected boolean transitionRun(final TraceBuilder traceBuilder) {
+  protected boolean transitionRun(final Context context) {
     State state;
     State newState;
     do {
       state = _stateRef.get();
       if (state.getType() != StateType.INIT) {
+        // prevent cross-plan task sharing if enabled
+        if (!ParSeqGlobalConfiguration.isAllowCrossPlanSharingEnabled() &&
+            _startPlanId != null && !_startPlanId.equals(context.getPlanId())) {
```

Comment:
This is not bulletproof because it might be the case that _startPlanId == null here.
Since State is CASed atomically the _startPlanId should belong to State class. Then it would be ensured that _startPlanId != null.

Reply:
That was my first attempt to put _startPlanId in State class, but then I realized that all those static final State instances  (INIT, RUN, DONE we used broadly in transition) are not quite useful anymore. Not sure about impact of removing all those static final State instance.

Reply:
Based on our discussion, I have put startPlanId as part of State class, removed all those PENDING, DONE, RUNNING static final State instance with only INIT left.  This will make State and startPlanId transition atomic. Thanks for pointing this out.